### PR TITLE
Enhance Wobject

### PR DESCRIPTION
**Pull request by ZhongYic00**

### CODE REVIEW

File: src/server/kernel/wbackend.cpp [hunk 1]
```diff
@@ -189,8 +189,10 @@ void WBackend::destroy(WServer *server)
     for (auto i : d->outputList)
         outputRemoved(i);
 
-    qDeleteAll(d->inputList);
-    qDeleteAll(d->outputList);
+    for(auto i : d->inputList)
```
Comment (zccrs) [inline src/server/kernel/wbackend.cpp]: `for (`

File: src/server/kernel/wglobal.cpp [hunk 2]
```diff
@@ -21,15 +28,97 @@ WObjectPrivate *WObjectPrivate::get(WObject *qq)
...
 WObjectPrivate::~WObjectPrivate()
 {
 
 }
 
-WObjectPrivate::WObjectPrivate(WObject *qq)
-    : q_ptr(qq)
+WWrapObject::WWrapObject(WWrapObjectPrivate &d, QObject *parent)
+    : WObject(d, nullptr),
+    QObject(parent)
```
Comment (zccrs) [inline src/server/kernel/wglobal.cpp]: `, QObject(parent)`

File: src/server/kernel/wglobal.cpp [hunk 3]
```diff
@@ -21,15 +28,97 @@ WObjectPrivate *WObjectPrivate::get(WObject *qq)
...
+{
+    W_D(WWrapObject);
+    int index = d->connections.indexOf(connection);
+    if (index < 0)
+        return false;
+    d->connections.removeAt(index);
+    return QObject::disconnect(connection);
+}
+
+void WWrapObject::safeDelete()
```
Comment (zccrs) [inline src/server/kernel/wglobal.cpp]: unnecessary

File: src/server/kernel/wsurface.cpp [hunk 4]
```diff
@@ -191,7 +194,7 @@ WSurface *WSurfacePrivate::ensureSubsurface(wlr_subsurface *subsurface)
 
     auto qwsurface = QWSurface::from(subsurface->surface);
     auto surface = new WSurface(qwsurface, q_func());
-    QObject::connect(qwsurface, &QWSurface::beforeDestroy, surface, &WSurface::deleteLater);
+    WWrapObject::safeConnect(surface, &QWSurface::beforeDestroy, surface, [surface]{ surface->safeDeleteLater(); });
```
Comment (zccrs) [inline src/server/kernel/wsurface.cpp]: `WWrapObject::safeConnect(surface, &QWSurface::beforeDestroy, surface,  &WSurface::safeDeleteLater);` ?

File: src/server/kernel/wxdgsurface.cpp [hunk 5]
```diff
@@ -146,13 +139,15 @@ void WXdgSurfacePrivate::connect()
...
-    QObject::connect(handle, &QWXdgSurface::configure, q, [this] (wlr_xdg_surface_configure *event) {
+    WWrapObject::safeConnect(q, &QWXdgSurface::configure, q, [this] (wlr_xdg_surface_configure *event) {
         on_configure(event);
     });
-    QObject::connect(handle, &QWXdgSurface::ackConfigure, q, [this] (wlr_xdg_surface_configure *event) {
+    WWrapObject::safeConnect(q, &QWXdgSurface::ackConfigure, q, [this] (wlr_xdg_surface_configure *event) {
         on_ack_configure(event);
     });
 
+    // TODO
```
Comment (zccrs) [inline src/server/kernel/wxdgsurface.cpp]: What's action of that TODO?

File: src/server/qtquick/private/wquickinputmethodv2.cpp [hunk 6]
```diff
@@ -54,33 +54,26 @@ void WQuickInputMethodManagerV2::create()
...
     WQuickInputPopupSurfaceV2Private(QWInputPopupSurfaceV2 *h, WQuickInputPopupSurfaceV2 *qq)
-        : WObjectPrivate(qq)
+        : WWrapObjectPrivate(qq)
         , handle(h)
         , popupSurface(WSurface::fromHandle(h->surface()))
     {
         if (!popupSurface) {
             popupSurface = new WSurface(h->surface());
-            QObject::connect(h->surface(), &QWSurface::beforeDestroy, popupSurface, &WSurface::deleteLater);
+            QObject::connect(h->surface(), &QWSurface::beforeDestroy, popupSurface, [this]{ popupSurface->safeDeleteLater(); });
```
Comment (zccrs) [inline src/server/qtquick/private/wquickinputmethodv2.cpp]: Can `&WSurface::safeDeleteLater`?

File: src/server/qtquick/private/wquickxwayland.cpp [hunk 7]
```diff
@@ -64,11 +64,11 @@ void XWayland::surfaceAdded(WXWaylandSurface *surface)
 
     QObject::connect(surface, &WXWaylandSurface::isToplevelChanged,
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxwayland.cpp]: Needs safeConnect too?

File: src/server/qtquick/woutputhelper.cpp [hunk 8]
```diff
@@ -48,14 +48,14 @@ class WOutputHelperPrivate : public WObjectPrivate
...
-        QObject::connect(qwoutput(), &QWOutput::needsFrame, qq, [this] {
+        WWrapObject::safeConnect(output, &QWOutput::needsFrame, qq, [this] {
             setNeedsFrame(true);
             qwoutput()->QWOutput::scheduleFrame();
         });
-        QObject::connect(qwoutput(), &QWOutput::damage, qq, [this] {
+        WWrapObject::safeConnect(output, &QWOutput::damage, qq, [this] {
             on_damage();
         });
         QObject::connect(output, &WOutput::modeChanged, qq, [this] {
```
Comment (zccrs) [inline src/server/qtquick/woutputhelper.cpp]: `safeConnect` ?

File: src/server/qtquick/wsurfaceitem.cpp [hunk 9]
```diff
@@ -460,6 +453,7 @@ void WSurfaceItemContent::itemChange(ItemChange change, const ItemChangeData &da
 {
     QQuickItem::itemChange(change, data);
     W_D(WSurfaceItemContent);
+
```
Comment (zccrs) [inline src/server/qtquick/wsurfaceitem.cpp]: remove

File: src/server/qtquick/wsurfaceitem.cpp [hunk 10]
```diff
@@ -1050,8 +1039,8 @@ void WSurfaceItemPrivate::initForSurface()
 
     QObject::connect(surface->handle(), &QWSurface::beforeDestroy, q,
                      &WSurfaceItem::releaseResources, Qt::DirectConnection);
-    QObject::connect(surface, SIGNAL(hasSubsurfaceChanged()), q, SLOT(onHasSubsurfaceChanged()));
-    QObject::connect(surface->handle(), &QWSurface::commit, q, &WSurfaceItem::onSurfaceCommit);
+    WWrapObject::safeConnect(surface, &WSurface::hasSubsurfaceChanged, q, [this]{ onHasSubsurfaceChanged(); });
```
Comment (zccrs) [inline src/server/qtquick/wsurfaceitem.cpp:1041]: Remove `W_PRIVATE_SLOT` in header file.

File: src/server/qtquick/wsurfaceitem.cpp [hunk 11]
```diff
@@ -1307,6 +1296,11 @@ qreal WSurfaceItemPrivate::getImplicitHeight() const
     return surfaceState->contentGeometry.height() + ps.height();
 }
 
+bool WSurfaceItem::isActive() const
```
Comment (zccrs) [inline src/server/qtquick/wsurfaceitem.cpp]: Rename to `isValid()`

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/wsurfaceitem.cpp (hunk 11 above)
Comment (ZhongYic00) [reply src/server/qtquick/wsurfaceitem.cpp]: QQuickItem is always valid, just looses surface update so inactive?

File: src/server/kernel/wxdgsurface.cpp (hunk 5 above)
Comment (ZhongYic00) [reply src/server/kernel/wxdgsurface.cpp]: toplevel not using safeconnect

**Comment by zccrs** [comment]
Move to https://github.com/vioken/waylib/pull/331

**Pull request closed without merging**